geoh5py/objects/surveys/direct_current.py

```diff
@@ -0,0 +1,282 @@
+        self._metadata: dict | None = None
+        self._ab_cell_id: ReferencedData | None = None
+
+        super().__init__(object_type, **kwargs)
+
+    @property
+    def ab_cell_id(self) -> ReferencedData | None:
+        """
+        Data object
+        """
+        if getattr(self, "_ab_cell_id", None) is None:
+            child = self.get_data("A-B Cell ID")
+            if any(child) and isinstance(child[0], ReferencedData):
+                self.ab_cell_id = child[0]
+
+        if getattr(self, "_ab_cell_id", None) is not None:
+            return self._ab_cell_id
+
+        return None
+
+    @ab_cell_id.setter
+    def ab_cell_id(self, data: Data | np.ndarray):
+
+        if isinstance(data, Data):
+            assert isinstance(
+                data, ReferencedData
+            ), f"ab_cell_id must be of type {ReferencedData}"
+
+            self._ab_cell_id = data
+
+        else:
+            assert data.dtype == "int32", "ab_cell_id must be of type 'int32'"
+            if any(self.get_data("A-B Cell ID")):
+                child = self.get_data("A-B Cell ID")[0]
+                if isinstance(child, ReferencedData):
+                    child.values = data
+            else:
+                if (
+                    getattr(self, "current_electrodes", None) is not None
+                    and getattr(self.current_electrodes, "ab_cell_id", None) is not None
+                ):
+                    entity_type = self.current_electrodes.ab_cell_id.entity_type
+                else:
+                    entity_type = {"primitive_type": "REFERENCED"}
+
+                data = self.add_data(
+                    {
+                        "A-B Cell ID": {
+                            "values": data,
+                            "association": "CELL",
+                            "entity_type": entity_type,
+                        }
+                    }
+                )
+
+                if isinstance(data, ReferencedData):
+                    self._ab_cell_id = data
+
+        self.workspace.finalize()
+
+    @property
+    def ab_map(self) -> dict | None:
+        """
+        Get the ReferenceData.value_map of the ab_value_id
+        """
+        if isinstance(self.ab_cell_id, ReferencedData):
+            return self.ab_cell_id.value_map
+        return None
+
+    @property
+    def current_electrodes(self):
+        """
+        The associated current electrode object (sources).
+        """
+        assert self.metadata is not None, "No Current-Receiver metadata set."
+        currents = self.metadata["Current Electrodes"]
+
+        try:
+            return self.workspace.get_entity(currents)[0]
+        except IndexError:
+            print("Associated CurrentElectrode entity not found in Workspace.")
+            return None
+
+    @current_electrodes.setter
+    def current_electrodes(self, current_electrodes: CurrentElectrode):
+
+        assert isinstance(current_electrodes, CurrentElectrode), (
```

Comment:
I think we should try to stick to try/except raise blocks or if condition raise in production code to raise custom errors.  I know they're more verbose, but assertions a more meant for debugging and testing.

Reply:
Done